Deadlock found when trying to get lock; try restarting transaction

# Bug Report

**Current behavior**

Running nested mutations in quick succession results in an internal server error, and this cluster log:

```
prisma-database_1  | java.sql.SQLTransactionRollbackException: (conn=164) Deadlock found when trying to get lock; try restarting transaction
prisma-database_1  |    at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:165)
prisma-database_1  |    at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:106)
prisma-database_1  |    at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:235)
prisma-database_1  |    at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:224)
prisma-database_1  |    at org.mariadb.jdbc.MariaDbPreparedStatementClient.execute(MariaDbPreparedStatementClient.java:159)
prisma-database_1  |    at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
prisma-database_1  |    at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
prisma-database_1  |    at slick.jdbc.StatementInvoker.results(StatementInvoker.scala:39)
prisma-database_1  |    at slick.jdbc.StatementInvoker.iteratorTo(StatementInvoker.scala:22)
prisma-database_1  |    at slick.jdbc.Invoker.first(Invoker.scala:31)
prisma-database_1  |    at slick.jdbc.Invoker.first$(Invoker.scala:30)
prisma-database_1  |    at slick.jdbc.StatementInvoker.first(StatementInvoker.scala:16)
prisma-database_1  |    at slick.jdbc.StreamingInvokerAction$HeadAction.run(StreamingInvokerAction.scala:52)
prisma-database_1  |    at slick.jdbc.StreamingInvokerAction$HeadAction.run(StreamingInvokerAction.scala:51)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:240)
prisma-database_1  |    at scala.collection.Iterator.foreach(Iterator.scala:929)prisma-database_1  |    at scala.collection.Iterator.foreach$(Iterator.scala:929)
prisma-database_1  |    at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
prisma-database_1  |    at scala.collection.IterableLike.foreach(IterableLike.scala:71)
prisma-database_1  |    at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
prisma-database_1  |    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:240)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:238)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:240)
prisma-database_1  |    at scala.collection.Iterator.foreach(Iterator.scala:929)prisma-database_1  |    at scala.collection.Iterator.foreach$(Iterator.scala:929)
prisma-database_1  |    at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
prisma-database_1  |    at scala.collection.IterableLike.foreach(IterableLike.scala:71)
prisma-database_1  |    at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
prisma-database_1  |    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:240)
prisma-database_1  |    at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:238)
prisma-database_1  |    at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4(DBIOAction.scala:534)
prisma-database_1  |    at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4$adapted(DBIOAction.scala:534)
prisma-database_1  |    at scala.collection.Iterator.foreach(Iterator.scala:929)prisma-database_1  |    at scala.collection.Iterator.foreach$(Iterator.scala:929)
prisma-database_1  |    at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
prisma-database_1  | {"@timestamp":"2018-01-31T22:08:50.089+00:00","@version":1,"message":"{\"key\":\"request/new\",\"requestId\":\"api:api:cjd3mj44p0ati0120jbq
4k8xf\"}","logger_name":"com.prisma.api.server.ApiServer","thread_name":"single-server-akka.actor.default-dispatcher-4","level":"INFO","level_value":20000,"HOST
NAME":"95d5b2fb32e2"}
prisma-database_1  |    at scala.collection.IterableLike.foreach(IterableLike.scala:71)
prisma-database_1  |    at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
prisma-database_1  |    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
prisma-database_1  |    at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.run(DBIOAction.scala:534)
prisma-database_1  |    at slick.dbio.SynchronousDatabaseAction$$anon$11.run(DBIOAction.scala:571)
prisma-database_1  |    at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:240)
prisma-database_1  |    at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:240)
prisma-database_1  |    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
prisma-database_1  |    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
prisma-database_1  |    at java.lang.Thread.run(Thread.java:748)
prisma-database_1  | Caused by: java.sql.SQLException: Deadlock found when trying to get lock; try restarting transaction
prisma-database_1  | Query is: insert into `benchmark@dev`.`_ArticleToSource` (`id`, `A`, `B`)
prisma-database_1  |            Select 'cjd3mj44b0atf012047wa8gjf', (select id from `benchmark@dev`.`Article` where `id` = ?), `id`
prisma-database_1  |            FROM `benchmark@dev`.`Source` where `slug` = ? on duplicate key update `benchmark@dev`.`_ArticleToSource`.id=`benchmark@dev`.`_A
rticleToSource`.id, parameters ['cjd3mj44a0ata0120dwlzt1n2','test']
prisma-database_1  |    at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:146)
prisma-database_1  |    at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:217)
prisma-database_1  |    at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:218)
prisma-database_1  |    ... 43 more
```

**Reproduction**

I prepared a reproduction with an older version here: https://github.com/marktani/import-testing

**Expected behavior?**

The deadlock does not occur, or when a deadlock occurs, the transaction is being retried for a fixed number of times.

Having the same issue

`prisma-database_1  | Caused by: java.sql.SQLException: Deadlock found when trying to get lock; try restarting transaction
prisma-database_1  | Query is: insert into `prisma-airbnb-example@dev`.`_UserInvitationSent` (`id`, `A`, `B`) Select 'cjf1al37j02050797etcqygyu', (SELECT `id` FROM (SELECT * FROM `prisma-airbnb-example@dev`.`Invitation`) IDFROMWHERE WHERE `id` = ?) , `id` FROM `prisma-airbnb-example@dev`.`User` where `id` = ? on duplicate key update `prisma-airbnb-example@dev`.`_UserInvitationSent`.id = `prisma-airbnb-example@dev`.`_UserInvitationSent`.id, parameters ['cjf1al37f020107977czzcu67','cjex5uaoo013z0797b1noyecm']`

--- a/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala
+++ b/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala
@@ -227,13 +227,14 @@ object DatabaseMutationBuilder {
   // region HELPERS
 
   def idFromWhere(projectId: String, where: NodeSelector): SQLActionBuilder = {
-    sql"(SELECT `id` FROM (SELECT * FROM `#$projectId`.`#${where.model.name}`) IDFROMWHERE WHERE `#${where.field.name}` = ${where.fieldValue})"
+    if (where.isId) {
+      sql"${where.fieldValue}"
+    } else {
+      sql"(SELECT `id` FROM (SELECT * FROM `#$projectId`.`#${where.model.name}`) IDFROMWHERE WHERE `#${where.field.name}` = ${where.fieldValue})"
+    }
   }
 
-  def idFromWhereEquals(projectId: String, where: NodeSelector): SQLActionBuilder = where.isId match {
-    case true  => sql" = ${where.fieldValue}"
-    case false => sql" = " ++ idFromWhere(projectId, where)
-  }
+  def idFromWhereEquals(projectId: String, where: NodeSelector): SQLActionBuilder = sql" = " ++ idFromWhere(projectId, where)
 
   def idFromWherePath(projectId: String, where: NodeSelector): SQLActionBuilder = {
     sql"(SELECT `id` FROM (SELECT  * From `#$projectId`.`#${where.model.name}`) IDFROMWHEREPATH WHERE `#${where.field.name}` = ${where.fieldValue})"

--- a/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala
+++ b/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala
@@ -174,8 +174,9 @@ object DatabaseMutationBuilder {
   }
 
   def cascadingDeleteChildActions(projectId: String, path: Path) = {
-    val deleteRelayIds  = (sql"DELETE FROM `#$projectId`.`_RelayId` WHERE `id` IN " ++ pathQueryForLastChild(projectId, path)).asUpdate
-    val deleteDataItems = (sql"DELETE FROM `#$projectId`.`#${path.lastModel.name}` WHERE `id` IN " ++ pathQueryForLastChild(projectId, path)).asUpdate
+    val deleteRelayIds = (sql"DELETE FROM `#$projectId`.`_RelayId` WHERE `id` IN (" ++ pathQueryForLastChild(projectId, path) ++ sql")").asUpdate
+    val deleteDataItems =
+      (sql"DELETE FROM `#$projectId`.`#${path.lastModel.name}` WHERE `id` IN (" ++ pathQueryForLastChild(projectId, path) ++ sql")").asUpdate
     DBIO.seq(deleteRelayIds, deleteDataItems)
   }
 
@@ -267,7 +268,7 @@ object DatabaseMutationBuilder {
 
         sql"(SELECT `#${last.childRelationSide}`" ++
           sql" FROM (SELECT * FROM `#$projectId`.`#${last.relation.id}`) PATHQUERY" ++
-          sql" WHERE " ++ childWhere ++ sql"`#${last.parentRelationSide}` IN " ++ pathQueryForLastParent(projectId, path) ++ sql")"
+          sql" WHERE " ++ childWhere ++ sql"`#${last.parentRelationSide}` IN (" ++ pathQueryForLastParent(projectId, path) ++ sql"))"
     }
   }
 
@@ -302,17 +303,17 @@ object DatabaseMutationBuilder {
 
   def oldParentFailureTrigger(project: Project, path: Path) = {
     val table = path.lastRelation_!.id
-    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDPARENTPATHFAILURETRIGGER WHERE `#${path.childSideOfLastEdge}` IN " ++ pathQueryForLastChild(
+    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDPARENTPATHFAILURETRIGGER WHERE `#${path.childSideOfLastEdge}` IN (" ++ pathQueryForLastChild(
       project.id,
-      path)
+      path) ++ sql")"
     triggerFailureWhenExists(project, query, table)
   }
 
   def oldParentFailureTriggerByField(project: Project, path: Path, field: Field) = {
     val table = field.relation.get.id
-    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDPARENTPATHFAILURETRIGGERBYFIELD WHERE `#${field.oppositeRelationSide.get}` IN " ++ pathQueryForLastChild(
+    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDPARENTPATHFAILURETRIGGERBYFIELD WHERE `#${field.oppositeRelationSide.get}` IN (" ++ pathQueryForLastChild(
       project.id,
-      path)
+      path) ++ sql")"
     triggerFailureWhenExists(project, query, table)
   }
 
@@ -329,9 +330,9 @@ object DatabaseMutationBuilder {
 
   def oldChildFailureTrigger(project: Project, path: Path) = {
     val table = path.lastRelation_!.id
-    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDCHILDPATHFAILURETRIGGER WHERE `#${path.parentSideOfLastEdge}` IN " ++ pathQueryForLastParent(
+    val query = sql"SELECT `id` FROM `#${project.id}`.`#$table` OLDCHILDPATHFAILURETRIGGER WHERE `#${path.parentSideOfLastEdge}` IN (" ++ pathQueryForLastParent(
       project.id,
-      path)
+      path) ++ sql")"
     triggerFailureWhenExists(project, query, table)
   }
 

--- a/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseQueryBuilder.scala
+++ b/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseQueryBuilder.scala
@@ -291,7 +291,11 @@ object DatabaseQueryBuilder {
   }
 
   def existsByPath(projectId: String, path: Path) = {
-    sql"select exists" ++ DatabaseMutationBuilder.pathQueryForLastChild(projectId, path)
+    if (path.edges.isEmpty && path.root.isId) {
+      sql"select exists (SELECT * FROM `#$projectId`.`#${path.root.model.name}` WHERE `id` = ${path.root.fieldValue})"
+    } else {
+      sql"select exists" ++ DatabaseMutationBuilder.pathQueryForLastChild(projectId, path)
+    }
   }
 
   def selectFromScalarList(projectId: String, modelName: String, fieldName: String, nodeIds: Vector[String]): SQLActionBuilder = {

--- a/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/impl/DatabaseMutactionExecutorImpl.scala
+++ b/server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/impl/DatabaseMutactionExecutorImpl.scala
@@ -3,6 +3,7 @@ package com.prisma.api.connector.mysql.impl
 import com.prisma.api.connector.mysql.DatabaseMutactionInterpreter
 import com.prisma.api.connector._
 import slick.jdbc.MySQLProfile.api._
+import slick.jdbc.TransactionIsolation
 
 import scala.concurrent.{ExecutionContext, Future}
 
@@ -20,7 +21,7 @@ case class DatabaseMutactionExecutorImpl(
       case false => DBIO.seq(interpreters.map(_.action): _*)
     }
     clientDb
-      .run(singleAction)
+      .run(singleAction.withTransactionIsolation(TransactionIsolation.ReadCommitted))
       .recover {
         case error =>
           val mappedError = combinedErrorMapper.lift(error).getOrElse(error)

--- a/server/servers/api/src/test/scala/com/prisma/api/mutations/DeadlockSpec.scala
+++ b/server/servers/api/src/test/scala/com/prisma/api/mutations/DeadlockSpec.scala
@@ -173,4 +173,50 @@ class DeadlockSpec extends FlatSpec with Matchers with ApiBaseSpec {
     Await.result(Future.traverse(0 to 50)((i) => exec(i)), Duration.Inf)
   }
 
+  "deleting many items" should "not cause deadlocks" in {
+    val project = SchemaDsl() { schema =>
+      val comment = schema.model("Comment").field("text", _.String)
+      schema.model("Todo").oneToManyRelation("comments", "todo", comment).field("a", _.String)
+    }
+    database.setup(project)
+
+    def create() =
+      Future(
+        server.query(
+          """mutation {
+        |  createTodo(
+        |    data: {
+        |      a: "b",
+        |      comments: {
+        |        create: [{text: "comment1"}, {text: "comment2"}]
+        |      }
+        |    }
+        |  ){
+        |    id
+        |    comments { id }
+        |  }
+        |}""".stripMargin,
+          project
+        ))
+
+    val todoIds = Await.result(Future.traverse(0 to 50)((i) => create()), Duration.Inf).map(_.pathAsString("data.createTodo.id"))
+
+    def exec(id: String) =
+      Future(
+        server.query(
+          s"""mutation {
+             |  deleteTodo(
+             |    where: { id: "${id}" }
+             |  ){
+             |    a
+             |  }
+             |}
+      """.stripMargin,
+          project
+        )
+      )
+
+    Await.result(Future.traverse(todoIds)((i) => exec(i)), Duration.Inf)
+  }
+
 }

--- a/server/servers/api/src/test/scala/com/prisma/api/mutations/DeadlockSpec.scala
+++ b/server/servers/api/src/test/scala/com/prisma/api/mutations/DeadlockSpec.scala
@@ -1,0 +1,176 @@
+package com.prisma.api.mutations
+
+import java.util.concurrent.Executors
+
+import com.prisma.api.ApiBaseSpec
+import com.prisma.shared.schema_dsl.SchemaDsl
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class DeadlockSpec extends FlatSpec with Matchers with ApiBaseSpec {
+
+  implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(100))
+
+  "creating many items" should "not cause deadlocks" in {
+    val project = SchemaDsl() { schema =>
+      val comment = schema.model("Comment").field("text", _.String)
+      schema.model("Todo").oneToManyRelation("comments", "todo", comment).field("a", _.String)
+    }
+    database.setup(project)
+
+    implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(100))
+
+    def exec(i: Int) =
+      Future(
+        server.query(
+          s"""mutation {
+             |  createTodo(
+             |    data:{
+             |      a: "a"
+             |    }
+             |  ){
+             |    a
+             |  }
+             |}
+      """.stripMargin,
+          project
+        )
+      )
+
+    Await.result(Future.traverse(0 to 50)((i) => exec(i)), Duration.Inf)
+  }
+
+  "updating single item many times" should "not cause deadlocks" in {
+    val project = SchemaDsl() { schema =>
+      val comment = schema.model("Comment").field("text", _.String)
+      schema.model("Todo").oneToManyRelation("comments", "todo", comment).field("a", _.String)
+    }
+    database.setup(project)
+
+    val createResult = server.query(
+      """mutation {
+        |  createTodo(
+        |    data: {
+        |      comments: {
+        |        create: [{text: "comment1"}, {text: "comment2"}]
+        |      }
+        |    }
+        |  ){
+        |    id
+        |    comments { id }
+        |  }
+        |}""".stripMargin,
+      project
+    )
+
+    val todoId     = createResult.pathAsString("data.createTodo.id")
+    val comment1Id = createResult.pathAsString("data.createTodo.comments.[0].id")
+    val comment2Id = createResult.pathAsString("data.createTodo.comments.[1].id")
+
+    def exec(i: Int) =
+      Future(
+        server.query(
+          s"""mutation {
+             |  updateTodo(
+             |    where: { id: "${todoId}" }
+             |    data:{
+             |      a: "${i}"
+             |    }
+             |  ){
+             |    a
+             |  }
+             |}
+      """.stripMargin,
+          project
+        )
+      )
+
+    Await.result(Future.traverse(0 to 50)((i) => exec(i)), Duration.Inf)
+  }
+
+  "updating single item and relations many times" should "not cause deadlocks" in {
+    val project = SchemaDsl() { schema =>
+      val comment = schema.model("Comment").field("text", _.String)
+      schema.model("Todo").oneToManyRelation("comments", "todo", comment).field("a", _.String)
+    }
+    database.setup(project)
+
+    val createResult = server.query(
+      """mutation {
+        |  createTodo(
+        |    data: {
+        |      comments: {
+        |        create: [{text: "comment1"}, {text: "comment2"}]
+        |      }
+        |    }
+        |  ){
+        |    id
+        |    comments { id }
+        |  }
+        |}""".stripMargin,
+      project
+    )
+
+    val todoId     = createResult.pathAsString("data.createTodo.id")
+    val comment1Id = createResult.pathAsString("data.createTodo.comments.[0].id")
+    val comment2Id = createResult.pathAsString("data.createTodo.comments.[1].id")
+
+    def exec(i: Int) =
+      Future(
+        server.query(
+          s"""mutation {
+             |  updateTodo(
+             |    where: { id: "${todoId}" }
+             |    data:{
+             |      a: "${i}"
+             |      comments: {
+             |        update: [{where: {id: "${comment1Id}"}, data: {text: "update ${i}"}}]
+             |      }
+             |    }
+             |  ){
+             |    a
+             |  }
+             |}
+      """.stripMargin,
+          project
+        )
+      )
+
+    Await.result(Future.traverse(0 to 50)((i) => exec(i)), Duration.Inf)
+  }
+
+  "creating many items with relations" should "not cause deadlocks" in {
+    val project = SchemaDsl() { schema =>
+      val comment = schema.model("Comment").field("text", _.String)
+      schema.model("Todo").oneToManyRelation("comments", "todo", comment).field("a", _.String)
+    }
+    database.setup(project)
+
+    def exec(i: Int) =
+      Future(
+        server.query(
+          s"""mutation {
+               |  createTodo(
+               |    data:{
+               |      a: "a",
+               |      comments: {
+               |        create: [
+               |           {text: "first comment: ${i}"}
+               |        ]
+               |      }
+               |    }
+               |  ){
+               |    a
+               |  }
+               |}
+      """.stripMargin,
+          project
+        )
+      )
+
+    Await.result(Future.traverse(0 to 50)((i) => exec(i)), Duration.Inf)
+  }
+
+}

